overlord/snapshotstate/backend/backend.go

```diff
@@ -476,9 +476,69 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 var ErrCannotCancel = errors.New("cannot cancel: import already finished")
 
+// XXX: move to a new errutil ?
+//
+// MultiError is a helper to handle mutliple errors.
+type MultiError struct {
+	header string
+	errs   []error
+}
+
+// NewMultiError returns a new MultiError struct initialized with
+// the given format string that explains what operation potentially
+// want wrong. MultiError can be nested and will render correctly
+// in these cases.
+//
+// Note that "Append" needs to get called to add actual errors.
+func NewMultiError(format string, v ...interface{}) *MultiError {
+	return &MultiError{header: fmt.Sprintf(format, v...)}
+}
+
+// Append appends an error to the MultiError.
+func (me *MultiError) Append(e error) {
+	me.errs = append(me.errs, e)
+}
+
+// Error formats the error string.
+func (me *MultiError) Error() string {
+	return me.nestedError(0, make(map[error]struct{}))
+}
+
+// Err returns nil if no errors where added to the MultiError helper
+// and otherwise it returns a MultiError.
+func (me *MultiError) Err() error {
+	if len(me.errs) == 0 {
+		return nil
+	}
+	return me
+}
+
+// helper to ensure formating of nested MultiErrors works.
+func (me *MultiError) nestedError(level int, seen map[error]struct{}) string {
+	indent := strings.Repeat(" ", level)
+	buf := bytes.NewBufferString(fmt.Sprintf("%s:\n", me.header))
+	for i, err := range me.errs {
+		switch v := err.(type) {
+		case *MultiError:
+			if _, ok := seen[err]; ok {
```

Comment:
Nice idea to support nesting like this. Do we already have cases where this is going to be used?

Reply:
it feels a bit YAGNI, it would be kind of hard to support in something that is really reusable

Reply:
actually, this is used here,  if we wanted to generalize this we would need to use an interface somehow